File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.py
import sys
import math
input = sys.stdin.readline

n = int(input())
if n==1:
    print(0)
    exit(0)

arr = [True]*(n+1)
arr[0] = False
arr[1] = False

for i in range(2, int(math.sqrt(n)) + 1):
    if arr[i]:
        for j in range(i * i, n + 1, i):
            arr[j] = False
prime_list = []
for i in range(2,n+1):
    if arr[i] == True:
        prime_list.append(i)

left ,right = 0,0
temp_sum = prime_list[0]
cnt = 0

while right < len(prime_list):
    if temp_sum >n:
        temp_sum -= prime_list[left]
        left+=1
    else:
        if temp_sum==n:
            cnt+=1
        right+=1
        if right == len(prime_list):
            break
        temp_sum += prime_list[right]
print(cnt)

File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        
        if(n==1){
            System.out.println(0);
            return;
        }

        boolean[]isPrime = new boolean[n+1];
        Arrays.fill(isPrime,true);
        isPrime[0] = isPrime[1]= false;

        for(int i = 2;i*i<=n;i++){
            if(isPrime[i]){
                for(int j = i*i;j<=n;j+=i){
                    isPrime[j] = false;
                }
            }
        }
        ArrayList<Integer> primeList = new ArrayList<>();
        for(int i = 2;i<=n;i++){
            if(isPrime[i]){
                primeList.add(i);
            }
        }
        int left  = 0,right = 0;
        int tempSum =  0;
        int cnt = 0;

        while(right<primeList.size()){
            tempSum += primeList.get(right);
            while(tempSum>n){
                tempSum -= primeList.get(left);
                left++;
            }
            if(tempSum==n){
                cnt++;
            }
            right++;
        }
        System.out.println(cnt);
        
    }     
}
    
